src/Form/AuthenticationForm.php

```diff
@@ -456,77 +501,82 @@ protected function createSuggestion(\Exception $exception, KeyInterface $key): M
-    $exception_text = preg_replace('/(.*mfa_token=)([^\&\r\n]+)(.*)/', '$1***mfa-token***$3', $exception_text);
-    $exception_text = preg_replace('/(.*password=)([^\&\r\n]+)(.*)/', '$1***password***$3', $exception_text);
-    $exception_text = preg_replace('/(Authorization: (Basic|Bearer) ).*/', '$1***credentials***', $exception_text);
-
-    $text = json_encode($credentials, JSON_PRETTY_PRINT) .
-      PHP_EOL .
-      json_encode($keys, JSON_PRETTY_PRINT) .
-      PHP_EOL .
-      json_encode($this->config('apigee_edge.client')->get(), JSON_PRETTY_PRINT) .
-      PHP_EOL .
+    // Sanitize exception text.
+    $exception_text = preg_replace([
+      '/(.*refresh_token=)([^\&\r\n]+)(.*)/',
+      '/(.*mfa_token=)([^\&\r\n]+)(.*)/',
+      '/(.*password=)([^\&\r\n]+)(.*)/',
+      '/(Authorization: (Basic|Bearer) ).*/',
+    ], [
+      '$1***refresh-token***$3',
+      '$1***mfa-token***$3',
+      '$1***password***$3',
+      '$1***credentials***',
+    ], (string) $exception);
+
+    // Filter out any private values from config.
+    $client_config = array_filter($this->config('apigee_edge.client')->get(), function ($key) {
+      return !is_string($key) || $key[0] !== '_';
+    }, ARRAY_FILTER_USE_KEY);
+
+    return json_encode($credentials, JSON_PRETTY_PRINT) . PHP_EOL .
+      json_encode($keys, JSON_PRETTY_PRINT) . PHP_EOL .
+      json_encode($client_config, JSON_PRETTY_PRINT) . PHP_EOL .
       $exception_text;
-
-    return $text;
   }
 
   /**
    * Pass form array to the AJAX callback.
    *
-   * @param array $form
+   * @param array &$form
    *   An associative array containing the structure of the form.
    * @param \Drupal\Core\Form\FormStateInterface $form_state
    *   Form state object.
    *
-   * @return array
+   * @return \Drupal\Core\Ajax\AjaxResponse
    *   The AJAX response.
+   *
+   * @throws \Exception
    */
-  public function ajaxCallback(array $form, FormStateInterface $form_state): array {
-    return $form;
+  public function ajaxCallback(array &$form, FormStateInterface $form_state): AjaxResponse {
+    $response = new AjaxResponse();
+    // Get any status messages so they can be rendered in the placeholder.
+    $status_messages = $this->renderer->render(StatusMessages::renderMessages());
+    // Clear any existing messages from the initial page load.
+    $response->addCommand(new ReplaceCommand('div.messages', ''));
+    $response->addCommand(new ReplaceCommand('#apigee-edge-auth-form-messages', "<div id=\"apigee-edge-auth-form-messages\">{$status_messages}</div>"));
+
+    // Replace the debug element (an empty wrapper if validation passes).
+    $response->addCommand(new ReplaceCommand(
+      '#apigee-edge-auth-form-debug-info',
+      $form['debug']['#access'] ? $this->renderer->render($form['debug']) : '<div id="apigee-edge-auth-form-debug-info" data-drupal-selector="edit-debug-placeholder"></div>'
```

Comment:
Could not you use the renderer service to render the original, empty form element if access is FALSE instead of returning this "custom" HTML markup?

Reply:
Hm, indeed, that is unfortunate. There must be a better way to render the element instead of fake its HTML output, but let's leave it as it is for now.